ipi/utils/mintools.py

```diff
@@ -892,6 +893,229 @@ def L_BFGS(x0, d0, fdf, qlist, glist, fdf0, big_step, tol, itmax, m, scale, k):
+    _, g0 = fdf0
+    g0 = g0.flatten()
+
+    # Nocedal's notation
+    B = hessian
+
+    # Calculate direction
+    # When the inverse itself is not needed, people recommend solve(), not inv().
+    info(" @DampedBFGS: sk = np.linalg.solve(B, -g0) ...", verbosity.debug)
+    info(
+        "              The code randomly crashes here with some versions of Numpy "
+        "based on OpenBLAS.\n"
+        "              If this happens, use Numpy based on MKL, e.g. from Anaconda.",
+        verbosity.debug,
+    )
+    info("Operands:", verbosity.debug)
+    info("%s,  %s" % (type(B), str(B.shape)), verbosity.debug)
+    info("%s,  %s" % (type(g0), str(g0.shape)), verbosity.debug)
+    sk = np.linalg.solve(B, -g0)
+    info(" @DampedBFGS: Calculated direction.", verbosity.debug)
+
+    # Cosine of the (f, dx) angle
+    quality = -np.dot(sk / np.linalg.norm(sk), g0 / np.linalg.norm(g0))
+    info(" @DampedBFGS: Direction quality: %.4f." % quality, verbosity.debug)
+
+    # I use maximal cartesian atomic displacement as a measure of step length
+    maxdispl = np.amax(np.linalg.norm(sk.reshape(-1, 3), axis=1))
+    info(" @DampedBFGS: big_step = %.6f" % big_step, verbosity.debug)
+    if maxdispl > big_step:
+        info(
+            " @DampedBFGS: maxdispl before scaling: %.6f bohr" % maxdispl,
+            verbosity.debug,
+        )
+        sk *= big_step / maxdispl
+
+    info(
+        " @DampedBFGS: maxdispl:                %.6f bohr"
+        % (np.amax(np.linalg.norm(sk.reshape(-1, 3), axis=1))),
+        verbosity.debug,
+    )
+
+    # Force call
+    _, g = fdf(x0 + sk.reshape(x0.shape))
+    g = g.flatten()
+    # coordinates CHECKED
+
+    # Update hessian
+    yk = g - g0
+    skyk = np.dot(sk, yk)
+
+    # Equation 18.15 in Nocedal
+    theta = 1.0
+    Bsk = np.dot(B, sk)
+    sBs = np.dot(sk, Bsk)
+    # Damped update if rhok isn't sufficiently positive
+    if skyk < 0.2 * sBs:
+        theta = (0.8 * sBs) / (sBs - skyk)
+        info(
+            " @DampedBFGS: damping update of the Hessian; "
+            "(direction dot d_gradient) is small. "
+            "theta := %.6f" % theta,
+            verbosity.debug,
+        )
+        yk = theta * yk + (1 - theta) * Bsk
+        skyk = np.dot(sk, yk)
+    else:
+        info(" @DampedBFGS: Update of the Hessian, no damping.", verbosity.debug)
+
+    info(" @DampedBFGS: (s_k dot y_k) before reciprocating: %e" % skyk, verbosity.debug)
+    try:
+        rhok = 1.0 / skyk
+    except:
+        warning(" @DampedBFGS: caught ZeroDivisionError in 1/skyk.", verbosity.high)
+        rhok = 1e5
+
+    # Compute BFGS term (eq. 18.16 in Nocedal)
+    B += np.outer(yk, yk) * rhok - np.outer(Bsk, Bsk) / sBs
+
+    print("det(Hessian): %f" % np.linalg.det(B))
+    print("Condition number: %f" % np.linalg.cond(B))
+    eigvals = np.real(np.linalg.eigvals(B))
+    print("Positive definite: %r" % np.all(eigvals > 0))
+    print("Eigenvalues of the Hessian:")
```

Comment:
Print this only in debug mode

Reply:
I think this can be deleted completely. I just forgot to delete this, but it doesn't help at all unless you spend really long time thinking about all this stuff, have tools to visualize eigenvectors etc.